Improve CUB device-scope documentation (#1862)
NVIDIA/cccl: docs/cub/developer_overview.rst

Overview
====================================

Device-scope functionality is provided by classes called ``DeviceAlgorithm``,
where ``Algorithm`` is the implemented algorithm.
These classes then contain static member functions providing corresponding API entry points.

.. code-block:: c++

    struct DeviceAlgorithm {
      template <typename ...>
      CUB_RUNTIME_FUNCTION static cudaError_t Algorithm(
          void *d_temp_storage, size_t &temp_storage_bytes, ..., cudaStream_t stream = 0) {
        // optional: minimal argument checking or setup to call dispatch layer
        return DispatchAlgorithm<...>::Dispatch(d_temp_storage, temp_storage_bytes, ..., stream);
      }
    };

> Device-scope facilities always return ``cudaError_t`` and accept ``stream`` as the last parameter (main stream by default)

Reviewer: suggestion: I think *NULL stream* is the right term
```suggestion
Device-scope facilities always return ``cudaError_t`` and accept ``stream`` as the last parameter (NULL stream by default)
```

> Device-scope facilities always return ``cudaError_t`` and accept ``stream`` as the last parameter (main stream by default)

Reviewer: suggestion: maybe a concrete example could help:
```suggestion
For example, device-level reduce will look like `cub::DeviceReduce::Sum`.
Device-scope facilities always return ``cudaError_t`` and accept ``stream`` as the last parameter (main stream by default)
```

    void kernel(...) { // (4)
      using policy = ChainedPolicy::ActivePolicy; // selects policy of active device compilation pass (5)
      using agent = AgentAlgorithm<policy>; // instantiates (6)
      agent a{...};
      a.Process(); // calls (7)
    }

    template <typename Policy>
    struct AlgorithmAgent {  // (6)
      void Process() { ... } // (7)
    };

Let's look at each of the building blocks closer.

The dispatch entry point is typically represented by a static member function called ``DispatchAlgorithm::Dispatch``
that constructs an object of type ``DispatchAlgorithm``, filling it with all arguments to run the algorithm,
and passes it to the ``ChainedPolicy::Invoke`` function:

.. code-block:: c++

    template <..., // algorithm specific compile-time parameters
              typename SelectedPolicy> // also called: PolicyHub
    struct DispatchAlgorithm : SelectedPolicy { // TODO(bgruber): I see no need for inheritance, can we remove it?
      CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE static
      cudaError_t Dispatch(void *d_temp_storage, size_t &temp_storage_bytes, ..., cudaStream stream) {

>         if (/* nothing no items to process */) return cudaSuccess;

Reviewer: suggestion:
```suggestion
        if (/* no items to process */) return cudaSuccess;
```

>         if (/* nothing no items to process */) return cudaSuccess;

Reviewer: important: the contract is that we set the temporary storage bytes. As written, user might provide uninitialized temporary storage bytes containing a large value, first phase returns success error code, user allocates temporary storage of undefined size and potentially gets OOM. I must've missed that while writing this overview. Let's fix this while we are here to something along the lines of:
```suggestion
        if (/* no items to process */) {
          if (d_temp_storage == nullptr) {
            temp_storage_bytes = 1;
          }
          return cudaSuccess;
        }
```

Author: I took it from the previous version of the docs, but I was thinking the same: we should set the `temp_storage_bytes`.

        int ptx_version   = 0;
        const cudaError_t error = CubDebug(PtxVersion(ptx_version));
        if (cudaSuccess != error)
        {
          return error;
        }
        using MaxPolicy = typename SelectedPolicy::MaxPolicy;
        DispatchAlgorithm dispatch(..., stream);
        return CubDebug(MaxPolicy::Invoke(ptx_version, dispatch));
      }
    };

While the dispatch layers of many algorithms are publicly accessible,
and were used because they expose additional performance knobs,
we no longer expose it publicly for new algorithms.

> For many algorithms, the dispatch layer is publicly accessible and used directly by users,

Reviewer: ```suggestion
For many legacy algorithms, the dispatch layer is publicly accessible and used directly by users,
```

since it often exposes additional performance knobs or configuration,
like choosing the index type or policies to use.
Exposing the dispatch layer also allowed users to tune algorithms for their use cases.

> In the newly added algorithms, the dispatch layer should not be exposed publicly anymore.

Reviewer: question: this seem to be said in the previous paragraph already. Am I missing something?

Note how the kernel is instantiated on ``MaxPolicy`` (2) while the kernel launch configuration uses ``ActivePolicy`` (1).
This is an important optimization to reduce compilation-time:

.. code-block:: c++

    template <typename ChainedPolicy /* ... */ >
    __launch_bounds__(ChainedPolicy::ActivePolicy::AlgorithmPolicy::BLOCK_THREADS) __CUB_DETAIL_KERNEL_ATTRIBUTES
    void kernel(...) {
      using policy = ChainedPolicy::ActivePolicy::AlgorithmPolicy;
      using agent = AgentAlgorithm<policy>;

      __shared__ typename agent::TempStorage temp_storage; // allocate static shared memory for agent

      agent a{temp_storage, ...};
      a.Process();
    }

The kernel gets compiled for each PTX version (``N`` many) that was provided to the compiler.
During each device pass,
``ChainedPolicy`` compares ``CUB_PTX_ARCH`` against the template parameter ``PolicyPtxVersion``
to select an ``ActivePolicy`` type.
During the host pass,
``Invoke`` is compiled for each architecture in the tuning list (``M`` many).
If we used ``ActivePolicy`` instead of ``MaxPolicy`` as a kernel template parameter,

> we will compile ``O(M*N)`` kernels instead of ``O(N)``.

Reviewer: ```suggestion
we would compile ``O(M*N)`` kernels instead of ``O(N)``.
```

Overview
====================================

Device-scope functionality is provided by classes called ``DeviceAlgorithm``,
where ``Algorithm`` is the implemented algorithm.
These classes then contain static member functions providing corresponding API entry points.

.. code-block:: c++

    struct DeviceAlgorithm {
      template <typename ...>
      CUB_RUNTIME_FUNCTION static cudaError_t Algorithm(
          void *d_temp_storage, size_t &temp_storage_bytes, ..., cudaStream_t stream = 0) {
        // optional: minimal argument checking or setup to call dispatch layer
        return DispatchAlgorithm<...>::Dispatch(d_temp_storage, temp_storage_bytes, ..., stream);
      }
    };

Device-scope facilities always return ``cudaError_t`` and accept ``stream`` as the last parameter (main stream by default)
and the first two parameters are always ``void *d_temp_storage, size_t &temp_storage_bytes``.
The implementation may consist of some minimal argument checking, but should forward as soon as possible to the dispatch layer.
Device-scope algorithms are implemented in files located in `cub/device/device_***.cuh`.

In general, the use of a CUB algorithm consists of two phases.

> During the first phase,
> temporary storage size is calculated and returned in ``size_t &temp_storage_bytes``.
> During the second phase,
> ``temp_storage_bytes`` of memory is expected to be allocated and ``d_temp_storage`` is expected to be the pointer to this memory.

Reviewer 2: Could we turn this into a numbered list?
```suggestion
  1. Temporary storage size is calculated and returned in ``size_t &temp_storage_bytes``.
  2. ``temp_storage_bytes`` of memory is expected to be allocated and ``d_temp_storage`` is expected to be the pointer to this memory.
```
